lib/rack/canonical_host.rb

```diff
@@ -15,11 +15,13 @@ def call(env)
       host = evaluate_host(env)
       redirect = Redirect.new(env, host, options)
 
-      if redirect.canonical?
-        app.call(env)
-      else
-        redirect.response
+      begin
+        return redirect.response unless redirect.canonical?
+      rescue Addressable::URI::InvalidURIError
+        return [400, { Rack::CONTENT_TYPE => "text/plain", Rack::CONTENT_LENGTH => "0" }, []]
```

Comment:
this is based on what Rack uses internally:

https://github.com/rack/rack/blob/dff6cfd249832d32ab190e6d20605bce0d6c702d/lib/rack/auth/abstract/handler.rb#L31-L37